File: Sources/Olaf/UI/Model/NetworkStats.swift
import Foundation

/// Summary statistics computed from the currently visible network records (pure computation — testable).
/// The viewer's "Statistics" screen produces this from the currently visible (filtered) list.
struct NetworkStats {

    let totalRequests: Int
    let failureCount: Int            // 4xx/5xx or transport error (excluding cancelled)
    let cancelledCount: Int
    let averageDurationMs: Int?
    let medianDurationMs: Int?
    let p95DurationMs: Int?
    let totalRequestBytes: Int
    let totalResponseBytes: Int
    /// HTTP methods — descending.
    let methodCounts: [(name: String, count: Int)]
    /// Status classes ("2xx".."5xx", "Error", "Cancelled") — fixed order, zeros omitted.
    let statusClassCounts: [(name: String, count: Int)]
    /// Hosts with the most requests — top 5.
    let hostCounts: [(name: String, count: Int)]
    /// Slowest requests — top 5 (path + duration).
    let slowest: [(path: String, durationMs: Int)]

    var failurePercent: Int {
        totalRequests > 0 ? Int((Double(failureCount) / Double(totalRequests) * 100).rounded()) : 0
    }

    static func compute(from entries: [LogEntry]) -> NetworkStats {
        let infos = entries.compactMap(NetworkLogInfo.init)

        var methodTally: [String: Int] = [:]
        var statusTally: [String: Int] = [:]
        var hostTally: [String: Int] = [:]
        var durations: [Int] = []
        var slow: [(path: String, durationMs: Int)] = []
        var failures = 0, cancelled = 0, reqBytes = 0, respBytes = 0

        for info in infos {
            methodTally[(info.method ?? "GET").uppercased(), default: 0] += 1
            statusTally[Self.statusClass(of: info), default: 0] += 1
            if !info.host.isEmpty { hostTally[info.host, default: 0] += 1 }
            if info.cancelled {
                cancelled += 1
            } else if info.isFailure {
                failures += 1
            }
            reqBytes += info.requestBytes ?? 0
            respBytes += info.responseBytes ?? 0
            if let ms = info.durationMs {
                durations.append(ms)
                slow.append((path: info.path, durationMs: ms))
            }
        }

        durations.sort()
        return NetworkStats(
            totalRequests: infos.count,
            failureCount: failures,
            cancelledCount: cancelled,
            averageDurationMs: durations.isEmpty
                ? nil : durations.reduce(0, +) / durations.count,
            medianDurationMs: durations.isEmpty
                ? nil : durations[durations.count / 2],
            p95DurationMs: durations.isEmpty
                ? nil : durations[min(durations.count - 1, Int(Double(durations.count) * 0.95))],
            totalRequestBytes: reqBytes,
            totalResponseBytes: respBytes,
            methodCounts: methodTally.sorted { $0.value > $1.value }.map { (name: $0.key, count: $0.value) },
            statusClassCounts: Self.statusOrder.compactMap { name in
                statusTally[name].map { (name: name, count: $0) }
            },
            hostCounts: Array(
                hostTally.sorted { $0.value > $1.value }.map { (name: $0.key, count: $0.value) }.prefix(5)
            ),
            slowest: Array(slow.sorted { $0.durationMs > $1.durationMs }.prefix(5))
        )
    }

    // MARK: - Classification

    private static let statusOrder = ["2xx", "3xx", "4xx", "5xx", "Error", "Cancelled"]

    private static func statusClass(of info: NetworkLogInfo) -> String {
        if info.cancelled { return "Cancelled" }
        guard let status = info.statusCode else { return "Error" }
        switch status {
        case 200..<300: return "2xx"
        case 300..<400: return "3xx"
        case 400..<500: return "4xx"
        case 500...: return "5xx"
        default: return "Error"
        }
    }
}

File: Android/olaf/src/main/java/com/olaf/ui/model/NetworkStats.kt
package com.olaf.ui.model

import com.olaf.LogEntry
import kotlin.math.roundToInt

/**
 * Summary statistics over the currently visible network records. A pure computation, so it is
 * directly testable; the viewer's statistics screen feeds it the filtered list.
 */
internal data class NetworkStats(
    val totalRequests: Int,
    /** 4xx/5xx or a transport failure — cancellations are counted separately. */
    val failureCount: Int,
    val cancelledCount: Int,
    val averageDurationMs: Long?,
    val medianDurationMs: Long?,
    val p95DurationMs: Long?,
    val totalRequestBytes: Long,
    val totalResponseBytes: Long,
    /** HTTP methods, most frequent first. */
    val methodCounts: List<Pair<String, Int>>,
    /** Status classes in a fixed order, with empty classes omitted. */
    val statusClassCounts: List<Pair<String, Int>>,
    /** The five busiest hosts. */
    val hostCounts: List<Pair<String, Int>>,
    /** The five slowest requests. */
    val slowest: List<Pair<String, Long>>
) {

    val failurePercent: Int
        get() = if (totalRequests > 0) {
            (failureCount.toDouble() / totalRequests * 100).roundToInt()
        } else {
            0
        }

    companion object {
        private val statusOrder = listOf("2xx", "3xx", "4xx", "5xx", "Error", "Cancelled")

        fun compute(entries: List<LogEntry>): NetworkStats {
            val infos = entries.mapNotNull(NetworkLogInfo::from)

            val methodTally = LinkedHashMap<String, Int>()
            val statusTally = LinkedHashMap<String, Int>()
            val hostTally = LinkedHashMap<String, Int>()
            val durations = mutableListOf<Long>()
            val slow = mutableListOf<Pair<String, Long>>()
            var failures = 0
            var cancelled = 0
            var requestBytes = 0L
            var responseBytes = 0L

            for (info in infos) {
                val method = (info.method ?: "GET").uppercase()
                methodTally[method] = (methodTally[method] ?: 0) + 1

                val statusClass = statusClassOf(info)
                statusTally[statusClass] = (statusTally[statusClass] ?: 0) + 1

                if (info.host.isNotEmpty()) hostTally[info.host] = (hostTally[info.host] ?: 0) + 1

                when {
                    info.cancelled -> cancelled++
                    info.isFailure -> failures++
                }

                requestBytes += info.requestBytes ?: 0
                responseBytes += info.responseBytes ?: 0

                info.durationMs?.let {
                    durations.add(it)
                    slow.add(info.path to it)
                }
            }

            durations.sort()

            return NetworkStats(
                totalRequests = infos.size,
                failureCount = failures,
                cancelledCount = cancelled,
                averageDurationMs = durations.takeIf { it.isNotEmpty() }?.let { it.sum() / it.size },
                medianDurationMs = durations.getOrNull(durations.size / 2),
                p95DurationMs = durations.getOrNull(
                    minOf(durations.size - 1, (durations.size * 0.95).toInt()).coerceAtLeast(0)
                ),
                totalRequestBytes = requestBytes,
                totalResponseBytes = responseBytes,
                methodCounts = methodTally.entries.sortedByDescending { it.value }.map { it.key to it.value },
                statusClassCounts = statusOrder.mapNotNull { name ->
                    statusTally[name]?.let { name to it }
                },
                hostCounts = hostTally.entries.sortedByDescending { it.value }.take(5).map { it.key to it.value },
                slowest = slow.sortedByDescending { it.second }.take(5)
            )
        }

        private fun statusClassOf(info: NetworkLogInfo): String = when {
            info.cancelled -> "Cancelled"
            info.statusCode == null -> "Error"
            info.statusCode in 200..299 -> "2xx"
            info.statusCode in 300..399 -> "3xx"
            info.statusCode in 400..499 -> "4xx"
            info.statusCode >= 500 -> "5xx"
            else -> "Error"
        }
    }
}
